src/sparseml/modifiers/smoothquant/base.py

```diff
@@ -0,0 +1,172 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from dataclasses import dataclass
+from typing import Dict, Generic, List, Optional, Tuple, TypeVar
+
+from pydantic import Field
+
+from sparseml.core import Modifier
+from sparseml.core.model import ModifiableModel
+from sparseml.core.model.base import LT
+from sparseml.core.state import Event, State
+
+
+VT = TypeVar("VT")  # represents a generic vector
+
+__all__ = ["SmoothQuantScale", "SmoothQuantMapping", "SmoothQuantModifier"]
+
+
+@dataclass
+class SmoothQuantScale(Generic[VT]):
+    """
+    Dataclass for storing the channel-wise minimum and maximum values for a layer. This
+    is updated each forward pass during calibration
+
+    :param min_channel_vals: minimum output value seen so far, per channel
+    :param max_channel_vals: maximum output value seen so far, per channel
+    """
+
+    min_channel_vals: VT
+    max_channel_vals: VT
+
+
+@dataclass
+class SmoothQuantMapping(Generic[LT]):
+    """
+    Dataclass for storing the mapping between an activation layer and the following
+    weights that must be balanced during smoothing
+
+    :param smooth_name: name of the activation layer
+    :param smooth_layer: PyTorch module storing the activation layer
+    :param balance_layers: list of PyTorch modules that smooth_layer feeds into, must be
+    balanced to offset the smoothing of smooth_layer
+    """
+
+    smooth_name: str
+    smooth_layer: LT
+    balance_layers: List[LT]
+
+
+class SmoothQuantModifier(Modifier):
+    """
+    Implements the SmoothQuant algorithm from https://arxiv.org/abs/2211.10438. This
+    modifier performs a channel-wise smoothing of outliers in activations, making them
+    easier to quantize by reducing the dynamic range. The smoothing is offset by
+    applying the inverse operation to the next layer of weights, making the weights
+    slightly more difficult to quantize.
+
+    Because this modifier manipulates the weights of the model, it can only be used in
+    in one-shot and not during training. Activation ranges are determined by running a
+    small set of calibration data through the model.
+
+    :param smoothing_strength: alpha, intensity of smoothing to perform (0-1 range)
+    :param mappings: list activation layers to smooth, and the which layers to offset
+    the smoothing to for each activation
+    :param ignore: list of layers to ignore, even if they match a regex in mappings
+    :param logarithmic_equalization: Whether to use a logarithmic scale for smoothing
+    :param num_calibration_steps: number of samples to use for calibration, or None to
+    use the whole dataset
+    """
+
+    smoothing_strength: float = Field(..., alias="alpha")
+    mappings: List[Tuple]
+    ignore: Optional[List[str]] = None
+    logarithmic_equalization: Optional[bool] = False
```

Comment:
It seems that the logarithmic_equalization argument is not being used in this implementation. That's fine. I actually suggest that you remove this argument from SmoothQuantModifier.

The only difference between smoothquant and logarithmic equalization is how the smoothing scale is computed. I suggest that you move the smoothing scale computation into a separate method, and then we can implement LogarithmicEqualizationModifier inheriting from SmoothQuantModifier and just changing this method.

Reply:
BTW, LogarithmicEqualizationModifier can be implemented in a different PR